labeling.py

```diff
@@ -0,0 +1,155 @@
+# face detection and printing errors to index file
+
+import os
+import sys
+from shutil import copyfile
+
+import cv2
+
+
+def label_faces():
+    """
+    This function detects faces in images in a directory and then places the bounding boxes in a txt file in YOLO
+    format.
+    It also creates an index file with images where it could not detect a face or detected more than one face for later
+    manual labelling.
+    """
+    # sets the emotion
+    try:
+        emotion = sys.argv[1]
+    except IndexError:
+        emotion = 'happy'
+        print('Emotion set to {}'.format(emotion))
+
+    # sets the image folder
+    try:
+        img_folder = sys.argv[2]
+    except IndexError:
+        print('No image folder entered, exiting program')
+        sys.exit(1)
+    emotion_folder = os.path.join(img_folder, emotion)
+    # checks for an index and a log folder and creates it if it does not exist
+    index = open(os.path.join(emotion_folder, 'index.txt'), 'w+')
+    log = open(os.path.join(emotion_folder, 'log.txt'), 'w+')
+
+    try:  # sets class-id for labeling bounding boxes, default 0
+        class_id = sys.argv[3]
+    except IndexError:
+        class_id = 0
+
+    try:  # the cascade file includes weights for open-cv to detect faces
+        cascade_file = os.path.join(img_folder, 'haarcascade_frontalface_default.xml')
+    except OSError:
+        print('No cascade file detected, please download and place in directory before running')
+        sys.exit(1)
+
+    # cycles through each file in the directory
+    for img_file in os.listdir(emotion_folder):
+        if img_file.endswith(('.png', '.jpg')):
+            face_cascade = cv2.CascadeClassifier(cascade_file)  # face detection idk how it works
+            try:
+                img = cv2.imread(os.path.join(emotion_folder, img_file))
+            except OSError:
+                continue
+            gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
+            faces = face_cascade.detectMultiScale(gray, 1.3, 5)
+
+            label = open(os.path.join(emotion_folder, img_file.split('.')[0] + '.txt'), 'w')
+            for (x, y, w, h) in faces:  # TODO add eye checking to make recognition more accurate
+                img_h, img_w = img.shape[:2]
+                try:
+                    center_x = (x + w / 2) / img_w
+                    center_y = (y + h / 2) / img_h
+                    rel_w = w / img_w
+                    rel_h = h / img_h
+                except ZeroDivisionError:
+                    continue  # writes to the text file in YOlO format
+                label.write('{} {} {} {} {}'.format(class_id, center_x, center_y, rel_w, rel_h))
```

Comment:
when it errors out (at the except clause), some variables such as center_x will not be defined.